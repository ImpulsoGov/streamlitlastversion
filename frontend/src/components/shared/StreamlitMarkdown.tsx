/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement, ReactNode, Fragment, PureComponent } from "react"
import ReactMarkdown from "react-markdown"
// @ts-ignore
import htmlParser from "react-markdown/plugins/html-parser"
// @ts-ignore
import { BlockMath, InlineMath } from "react-katex"
// @ts-ignore
import RemarkMathPlugin from "remark-math"
// @ts-ignore
import RemarkEmoji from "remark-emoji"
import CodeBlock from "components/elements/CodeBlock/"

import "katex/dist/katex.min.css"

export interface Props {
  /**
   * The Markdown formatted text to render.
   */
  source: string

  /**
   * True if HTML is allowed in the source string. If this is false,
   * any HTML will be escaped in the output.
   */
  allowHTML: boolean
}

/**
 * Wraps the <ReactMarkdown> component to include our standard
 * renderers and AST plugins (for syntax highlighting, HTML support, etc).
 */
export class StreamlitMarkdown extends PureComponent<Props> {
  public componentDidCatch = (): void => {
    const { source } = this.props

    throw Object.assign(new Error(), {
      name: "Error parsing Markdown or HTML in this string",
      message: <p>{source}</p>,
      stack: null,
    })
  }

  public render = (): ReactNode => {
    const { source, allowHTML } = this.props

    const renderers = {
      code: CodeBlock,
      link: linkWithTargetBlank,
      linkReference: linkReferenceHasParens,
      inlineMath: (props: { value: string }): ReactElement => (
        <InlineMath>{props.value}</InlineMath>
      ),
      math: (props: { value: string }): ReactElement => (
        <BlockMath>{props.value}</BlockMath>
      ),
    }

    const plugins = [RemarkMathPlugin, RemarkEmoji]
    const astPlugins = allowHTML ? [htmlParser()] : []

    return (
      <ReactMarkdown
        source={source}
        escapeHtml={!allowHTML}
        astPlugins={astPlugins}
        plugins={plugins}
        renderers={renderers}
      />
    )
  }
}

interface LinkProps {
  children: ReactElement
  href: string
}

interface LinkReferenceProps {
  children: [ReactElement]
  href: string
}

// Using target="_blank" without rel="noopener noreferrer" is a security risk:
// see https://mathiasbynens.github.io/rel-noopener
export function linkWithTargetBlank(props: LinkProps): ReactElement {
  return (
    <a href={props.href} target="_blank" rel="noopener noreferrer">
      {props.children}
    </a>
  )
}

// Handle rendering a link through a reference, ex [text](href)
// Don't convert to a link if only `[text]` and missing `(href)`
export function linkReferenceHasParens(
  props: LinkReferenceProps
): ReactElement | null {
  const { href, children } = props

  if (!href) {
    return children.length ? (
      <Fragment>[{children[0].props.children}]</Fragment>
    ) : null
  }

  return (
    <a href={href} target="_blank" rel="noopener noreferrer">
      {children}
    </a>
  )
}
