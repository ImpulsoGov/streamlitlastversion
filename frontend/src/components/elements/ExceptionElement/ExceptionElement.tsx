/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactNode, PureComponent } from "react"
import classNames from "classnames"
import { Map as ImmutableMap } from "immutable"
import { StreamlitMarkdown } from "components/shared/StreamlitMarkdown"
import "./ExceptionElement.scss"

export interface Props {
  width: number
  element: ImmutableMap<string, any>
}

/**
 * Functional element representing formatted text.
 */
class ExceptionElement extends PureComponent<Props> {
  public render(): ReactNode {
    const { element, width } = this.props

    const type = element.get("type")
    const message = element.get("message")
    const stackTrace = element.get("stackTrace")
    const isWarning = element.get("isWarning")

    // Build the message display.
    // On the backend, we use the StreamlitException type for errors that
    // originate from inside Streamlit. These errors have Markdown-formatted
    // messages, and so we wrap those messages inside our Markdown renderer.
    let messageNode: ReactNode
    if (element.get("messageIsMarkdown")) {
      let markdown = `**${type}**`
      if (message) {
        markdown += `: ${message}`
      }
      messageNode = <StreamlitMarkdown source={markdown} allowHTML={false} />
    } else {
      messageNode = (
        <>
          <span className="type">{type}</span>
          {message != null ? `: ${message}` : null}
        </>
      )
    }

    // Build the stack trace display, if we got a stack trace.
    let stackTraceNode: ReactNode = null
    if (stackTrace && stackTrace.size > 0) {
      stackTraceNode = (
        <>
          <div className="stack-trace-title">Traceback:</div>
          <pre className="stack-trace">
            <code>
              {stackTrace.map((row: string, indx: string) => (
                <div className="stack-trace-row" key={indx}>
                  {row}
                </div>
              ))}
            </code>
          </pre>
        </>
      )
    }

    const wrapperClasses = classNames("alert", "exception", "stException", {
      "alert-danger": !isWarning,
      "alert-warning": isWarning,
    })

    return (
      <div className={wrapperClasses} style={{ width }}>
        <div className="message">{messageNode}</div>
        {stackTraceNode}
      </div>
    )
  }
}

export default ExceptionElement
