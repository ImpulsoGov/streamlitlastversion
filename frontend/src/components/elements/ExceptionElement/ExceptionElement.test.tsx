/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import { shallow } from "enzyme"
import { fromJS } from "immutable"

import { StreamlitMarkdown } from "components/shared/StreamlitMarkdown"
import ExceptionElement, { Props } from "./ExceptionElement"

const getProps = (elementProps: object = {}): Props => ({
  element: fromJS({
    stackTrace: ["step 1", "step 2", "step 3"],
    type: "RuntimeError",
    message: "This is an exception of type RuntimeError",
    messageIsMarkdown: false,
    ...elementProps,
  }),
  width: 0,
})

describe("ExceptionElement Element", () => {
  const props = getProps()
  const wrapper = shallow(<ExceptionElement {...props} />)

  it("renders without crashing", () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render the complete stack", () => {
    expect(wrapper.find(".stack-trace-title").text()).toBe("Traceback:")

    const traceRows = wrapper.find("code .stack-trace-row")
    expect(traceRows.length).toBe(3)

    traceRows.forEach((val, id) => {
      expect(val.prop("children")).toBe(`step ${id + 1}`)
    })
  })

  it("should render markdown when it has messageIsMarkdown", () => {
    const props = getProps({
      messageIsMarkdown: true,
    })
    const wrapper = shallow(<ExceptionElement {...props} />)

    expect(wrapper.find(StreamlitMarkdown).length).toBe(1)
    expect(wrapper.find(StreamlitMarkdown).props()).toMatchSnapshot()
  })

  it("should render if there's no message", () => {
    const props = getProps({
      message: null,
    })
    const wrapper = shallow(<ExceptionElement {...props} />)

    expect(wrapper.find("div .message").text()).toBe("RuntimeError")
  })
})
