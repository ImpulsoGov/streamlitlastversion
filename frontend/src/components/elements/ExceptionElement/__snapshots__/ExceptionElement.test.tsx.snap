// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExceptionElement Element renders without crashing 1`] = `"<div class=\\"alert exception stException alert-danger\\" style=\\"width:0\\"><div class=\\"message\\"><span class=\\"type\\">RuntimeError</span>: This is an exception of type RuntimeError</div><div class=\\"stack-trace-title\\">Traceback:</div><pre class=\\"stack-trace\\"><code><div class=\\"stack-trace-row\\">step 1</div><div class=\\"stack-trace-row\\">step 2</div><div class=\\"stack-trace-row\\">step 3</div></code></pre></div>"`;

exports[`ExceptionElement Element should render markdown when it has messageIsMarkdown 1`] = `
Object {
  "allowHTML": false,
  "source": "**RuntimeError**: This is an exception of type RuntimeError",
}
`;
